src/cressi_goa_parser.c

```diff
@@ -271,85 +620,177 @@ cressi_goa_parser_get_field (dc_parser_t *abstract, dc_field_type_t type, unsign
 			time += interval;
-			complete = 1;
+			complete = true;
+		} else if (type == DEPTH_FREE) {
+			depth = value & ((layout->version < 4) ? 0x7ffu : 0xfffu);
+			time += interval;
+			complete = true;
 		} else if (type == TEMPERATURE) {
 			temperature = value;
-			have_temperature = 1;
-		} else if (type == TIME) {
-			time += value;
+			have_temperature = true;
+		} else if (type == SURFACE) {
+			// SURFACE values are not given in the sample rate, but as seconds
+			if (divemode == FREEDIVE_ADV && (offset + 2 < size)) {
+				if (time)
+					t = interval;
+				if (SUBSURFACE_CORRECTION(interval, (time + t)))
+					t = 2;
+				time += t;
+				depth = 0;
+
+				// Time (seconds).
+				sample.time = time * cressi_goa_time_multiplier;
+				cressi_goa_parser_callback (DC_SAMPLE_TIME, &sample, &ctx);
+				// Depth (1/10 m).
+				sample.depth = 0.0;
+				cressi_goa_parser_callback (DC_SAMPLE_DEPTH, &sample, &ctx);
+			}
+			// One `time` unit equals 500ms, so we have to multiply the `value` with 2 in order to count seconds
+			time += ((value * 2) - t);
+			if (divemode == FREEDIVE_ADV && t) {
+				unsigned int correction = SUBSURFACE_CORRECTION(interval, time) ? 1 : 0;
+				// Time (seconds).
+				sample.time = (time - correction) * cressi_goa_time_multiplier;
+				cressi_goa_parser_callback (DC_SAMPLE_TIME, &sample, &ctx);
+				// Depth (1/10 m).
+				sample.depth = 0.0;
+				cressi_goa_parser_callback (DC_SAMPLE_DEPTH, &sample, &ctx);
+			}
 		}
 
-		if (complete) {
+		if (complete && SUBSURFACE_ACCEPTABLE(interval, time)) {
 			// Time (seconds).
-			sample.time = time * 1000;
-			if (callback) callback (DC_SAMPLE_TIME, &sample, userdata);
+			sample.time = time * cressi_goa_time_multiplier;
+			cressi_goa_parser_callback (DC_SAMPLE_TIME, &sample, &ctx);
 
 			// Temperature (1/10 °C).
 			if (have_temperature) {
 				sample.temperature = temperature / 10.0;
-				if (callback) callback (DC_SAMPLE_TEMPERATURE, &sample, userdata);
-				have_temperature = 0;
+				cressi_goa_parser_callback (DC_SAMPLE_TEMPERATURE, &sample, &ctx);
+				have_temperature = false;
 			}
 
 			// Depth (1/10 m).
 			sample.depth = depth / 10.0;
-			if (callback) callback (DC_SAMPLE_DEPTH, &sample, userdata);
+			cressi_goa_parser_callback (DC_SAMPLE_DEPTH, &sample, &ctx);
 
 			// Gas change
 			if (divemode == SCUBA || divemode == NITROX) {
 				if (gasmix != gasmix_previous) {
 					sample.gasmix = gasmix;
-					if (callback) callback (DC_SAMPLE_GASMIX, &sample, userdata);
+					cressi_goa_parser_callback (DC_SAMPLE_GASMIX, &sample, &ctx);
 					gasmix_previous = gasmix;
 				}
 			}
 
-			complete = 0;
+			complete = false;
 		}
 
 		offset += 2;
 	}
 
+	if (!ctx.last_depth_is_zero) {
+		dc_sample_value_t sample = {0};
+		sample.time = time * cressi_goa_time_multiplier;
+		cressi_goa_parser_callback (DC_SAMPLE_TIME, &sample, &ctx);
+		sample.depth = 0.0;
+		cressi_goa_parser_callback (DC_SAMPLE_DEPTH, &sample, &ctx);
+	}
+
```

Comment:
I'm not sure if that's 100% correct, since something could cause the DC to stop measuring while still in dive mode.

One could still argue that it's correct, since you very likely stopped diving before reading out your DC ...

I'm undecided and for the sake of better looking graphs I'd leave it in.

Reply:
Even it's not there, applications should not expect a zero depth sample at the end is always present. Other backends don't do this either.